cli/src/commands/fix.rs

```diff
@@ -48,106 +69,253 @@ pub(crate) fn cmd_fix(
-            if let Some(parent_paths) = commit_paths.get(parent_id) {
-                paths.extend_from_slice(parent_paths);
-            }
-        }
-        let parent_tree = commit.parent_tree(tx.repo())?;
+        // Fix paths that are different from all parents. This means that files in merge
+        // commits are only fixed if they have evil changes, or if a parent was fixed.
+        let mut paths_to_fix: HashSet<RepoPathBuf> = HashSet::new();
         let tree = commit.tree()?;
+        let parent_tree = commit.parent_tree(tx.repo())?;
         let mut diff_stream = parent_tree.diff_stream(&tree, &EverythingMatcher);
         async {
             while let Some((repo_path, diff)) = diff_stream.next().await {
                 let (_before, after) = diff?;
+                // Deleted files have no file content to fix, and they have no terms in "after",
+                // so we don't add any tool inputs for them. Conflicted files produce one tool
+                // input for each side of the conflict.
                 for term in after.into_iter().flatten() {
-                    if let TreeValue::File { id, executable: _ } = term {
-                        file_ids.insert((repo_path.clone(), id));
-                        paths.push(repo_path.clone());
+                    // We currently only support fixing the content of normal files, so we skip
+                    // directories and symlinks, and we ignore the executable bit.
+                    if let TreeValue::File {
+                        id: _,
+                        executable: _,
+                    } = term
+                    {
+                        // TODO: Consider filename arguments and tool configuration instead of
+                        // passing every changed file into the tool. Otherwise, the tool has to
+                        // be modified to implement that kind of stuff.
+                        paths_to_fix.insert(repo_path.clone());
                     }
                 }
             }
             Ok::<(), BackendError>(())
         }
         .block_on()?;
-        commit_paths.insert(commit.id().clone(), paths);
+
+        // Also fix all paths that were fixed in ancestors, so we don't lose those
+        // changes. We do this instead of rebasing onto those changes.
+        for parent_id in commit.parent_ids() {
+            if let Some(parent_tool_inputs) = commit_tool_inputs.get(parent_id) {
+                for repo_path in parent_tool_inputs.keys() {
+                    paths_to_fix.insert(repo_path.clone());
+                }
+            }
+        }
+
+        // Get this commit's file IDs for the paths to be fixed. Conflicting files will
+        // have multiple versions to be fixed, so we map paths to a set of inputs.
+        let mut tool_inputs_map: HashMap<RepoPathBuf, Vec<ToolInput>> = HashMap::new();
+        for repo_path in paths_to_fix {
+            let value = tree.path_value(&repo_path);
```

Comment:
It might not matter in practice since they're cached, but it seems unfortunate to look up the trees and iterate over terms again here. I think I would prefer to populate `commit_tool_inputs` inside the `diff_stream` loop. Before that, where the old loop over parents was, copy over the data for parents from `commit_tool_inputs` (if there was an entry for the parent).

Reply:
This came about due to a sequence of refactorings... and it was very unnecessary :)